examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+  (define (parse input-string)
+    "Parses Brainfold input into a lexical tree."
+    (let ((match-cmds
+            (fn (str-list)
+              (match str-list
+                ((Nil) Nil)
+                ((Cons ">" _)
+                 (Cons BFRight (match-cmds (list:cdr str-list))))
+                ((Cons "<" _)
+                 (Cons BFLeft (match-cmds (list:cdr str-list))))
+                ((Cons "+" _)
+                 (Cons BFPlus (match-cmds (list:cdr str-list))))
+                ((Cons "-" _)
+                 (Cons BFMinus (match-cmds (list:cdr str-list))))
+                ((Cons "." _)
+                 (Cons BFPrint (match-cmds (list:cdr str-list))))
+                ((Cons "," _)
+                 (Cons BFInput (match-cmds (list:cdr str-list))))
+                ((Cons "[" _)
+                 (Cons (BFLoop (match-cmds (list:cdr str-list)))
+                       (match-cmds
+                        (list:nth-cdr
+                         (index-after-loop (list:cdr str-list))
+                         str-list))))
+                ((Cons "]" _) nil)
+                (_ (match-cmds (list:cdr str-list)))
+                ))))
+      (match-cmds (split-string input-string)))))
+
+
+;;;
+;;; Evaluation/Compilation
+;;;
+
+(coalton-toplevel
+
+  (define (exec cmd)
+    "Executes a given bf command string."
+    (match cmd
+      ((BFRight) (move-right))
+      ((BFLeft) (move-left))
+      ((BFPlus) (fromint (incr)))
+      ((BFMinus) (fromint (decr)))
+      ((BFPrint) (fromint (print-current-cell)))
+      ((BFInput) (fromint (take-input)))
+      ((BFLoop xs) (exec-loop xs))))
+
+  (define (exec-loop looped-cmds)
+    "Executes a loop of cmds, exiting when the value at the pointer is 0."
+    (let ((execute-loop-cmds
+            (fn (looped-cmds)
+              (if (list:null? looped-cmds)
+                  (fromint (current-bf-cell-value))
+                  (progn (exec (list:car looped-cmds))
+                         (execute-loop-cmds
+                          (list:cdr looped-cmds)))))))
+      (if (arith:zero? (current-bf-cell-value))
+          (fromint (current-bf-cell-value))
+          (progn (execute-loop-cmds looped-cmds)
+                 (exec-loop looped-cmds)))))
+
+  (define (eval input-string)
+    "Parses and evaluates a string of brainfold input."
+    (let ((execute-cmds
+            (fn (cmd-list)
+              (if (list:null? cmd-list)
+                  bf-array
+                  (progn (exec (list:car cmd-list))
+                         (execute-cmds
+                          (list:cdr cmd-list)))))))
+      (execute-cmds (parse input-string)))))
+
+;;;
+;;; Top Level
+;;;
+
+(coalton-toplevel
+
+  (declare run-program (String -> String))
+  (define (run-program bf-string)
+    "Takes a bf-command string and evaluates and executes its components
+after restarting the environment."
+    (reset-bf)
+    (eval bf-string)
+    (if (== "" (cell:read print-buffer))
```

Comment:
I'm not sure what the value of this `if` is. Just to say `Compiled` if nothing was ever written?

I'd get rid of it.